tests/usm/usm_api.h

```diff
@@ -251,6 +251,13 @@ class copy : public copyGeneric<T, count, sourceAllocation> {
    */
   static constexpr bool has_non_usm_support() { return true; }
 
+  /** @brief This test only works on devices with `sourceAllocation` support
+   */
+  static bool supports_device(sycl_cts::util::logger& log,
+                              const sycl::queue& queue) {
+    return check_device_support<sourceAllocation>(log, queue);
```

Comment:
That looks good.
Just thinking we could refactor globally the tests to have kinds of traits to express this at the class level to avoid all this copy paste of `supports_device()`.
Such as instead of supports_device
```c++
class copy : public copyGeneric<T, count, sourceAllocation> {
```
we could have just
```c++
class copy : public copyGeneric<T, count, sourceAllocation>, public supports_device<sourceAllocation> {
```

Reply:
Sorry, hit merge too early. @RossBrunton, wound you mind applying this suggestion, please?

Reply:
Sure, I can create a follow up merge request for this.

Unfortunately I've not really worked with traits much in C++, and am a bit unsure on what you are suggesting here. What would a definition of `supports_device` look like?

Reply:
I can only guess, but one way I see it can be done is by defining two class templates: "no_check" and "sourceAllocation_check". Both of them define `supports_device` method according to their names. `copy` and `memcpy` inherit from "sourceAllocation_check" template. All other classes from "no_check" template.
@keryell is that what you have in mind?

Reply:
Yes, this is the idea.